README.md
# PlanBranch
Manage your adult support center's activities.

## Features
* Set permissions and levels for residents.
* Assign training, counseling, and volunteer work for residents.
* Generate a space-efficient one-page schedule for the week.
* Develop week overview.
* Copy from week to day (then add more details on day roster,
  such as activities).

## Known Issues
### Outliers on destination schedule
* testing: newcomer & every 6 wks
* small class on Tues, Wed, Fri

### Not yet implemented
* Job Hunting on Days Specified (copied to day schedule)
* [ ] Override background color with ScheduleItemsReasons.SIR_Color if ScheduleItems.SIR_Id is set
  - only if the related ScheduleItemsReasons.SIR_WarnOnChange is set (to 1)
* [ ] See who can go to lighthouse on a certain day
* Add PeopleUnavailableReasons
* If People.Note is not null, Show ID in red and click/hover to see note.
* Italicize ShowAs string if related field in AT_Italics is 1
* Enforce SC_LimitPerWeek (prevent count of this type
  `> SC_LimitPerWeek`).
* Enforce SC_FractionOfDay (prevent > 1.0).
* Prefill the ScheduleItems.SIR_Id if ActivitiesTypes.SIR_Id is not null.
* Prefill the ScheduleItems.ET_Id if SlotCodes.ET_Id is not null.
* Fill entire row of schedule with reasons if a row in
  PeopleUnavailableReasons exists for the row in People.
* If WorkPermissionsTypes.PFT_Id is not null, that PeopleFlagType
  prevents scheduling the person for that permission.
* Manually track current value of ID of each table (Sample data is
  dependent on manually-assigned primary keys).
* Make sticky setting (0 or 1) for ScheduleItems that should be
  copied to next week (or possibly to a separate template table
  containing malleable default values).
* App-Defined Roles:
  - Master Scheduler (Jason)
  - Desk Scheduler
  - Counseling Scheduler
* The director or assistant director could call and place a resident in
  a staff position or direct that the resident must be working that day
  (see `higher level decision` in ScheduleItemsReasons).
* Teacher (only day classes affect master scheduler)
  - assigned to certain students
  - no truck, nor lighthouse that day
* Tutor: Tutor may ask to switch time later
  - Tutor says when they can come.
    - The master scheduler may switch which tutor a resident sees, such
      as by switching two residents.
* Every 4-5 weeks, someone gets overnight desk--usually give them off
  the next day

### Master Schedule Caveats Not Yet Implemented
* truck/cashier: (not guys flagged purple)
  - truck: only if healthy, & only certain guys
  - lighthouse: offsite thrift store w/ limited staff, 30min bus, 30 m
    walk (2 or 3 guys usually)
* Sundays off usually, else Sat

### Desk Schedule Features Not Yet Implemented
* Volunteers are at the desk when staff is not available
* Printed for each student from MS Access,
  but only item from day classes is GED class,
  so it doesn't need input from master day schedule
  - Access DB needs work: such as, hide fields on report
* Someone sends this to the master scheduler (to Jason).
* The master scheduler changes the timeslot.
  - same for Counselor schedule
    - REQUIRED: keep person w/ same counselor
      but prioritize tutoring because tutors are only available certain
      times (keeping residents with same tutor is preferred)

### Desk Schedule Caveats Not Yet Implemented
* Avoid revisiting each day roster multiple times (see related info that
  affects schedule)
* Groups don't always start the same week, but a group lasts 4 weeks.
* Tutoring schedule: Accommodate changes to tutor & resident
  availability.
  - Master scheduler (Jason) may sub, possibly at a different
    time.
* Discretionary or last-minute changes to day schedule (must go back to
  week schedule--posted week schedule must be changed by hand)
  - OR edit week schedule
* Day & Deskmen Schedule (keep record)
  - such as, switch.
    - must get permission to make further changes
* Staff members like to have a paper copy, but it has limited info:
  - day off, truck, lighthouse, kitchen, desk, work
    (class schedule is computerized, but there is a printable version)



## Master Day Schedule
("At a Glance")
* write-in codes usually done that day
* hidden column: lighthouse permission (darkorange)
  - [ ] change to colored bullet by name instead of hiding
    People with kitchen or truck permission can already go to lighthouse
    generally. It was just hidden because cumbersome to show multiple
    color codes in a spreadsheet program.

### Slots
Denote on master schedule as:
```HTML
Black numbers: tutoring time
<br/><span style="color:red">Red</span> numbers: counseling time
<br/><span style="background-color:plum">^^ 9am Computer class (W); ^ 4pm GED (T), Communication (W), or Discipleship (F) class</span>
```

### Slot Codes
* Black #: tutoring time (hour of day)
* Red #: counseling time 1/wk (hour of day)
* `OFF*`: personal appt on 2nd off day (not movable)
  - otherwise 2nd off day can be any day (movable)
* `OFF` (no asterisk): split 2nd off day, such as if they have an
  afternoon personal appt. (can have 2 half days off instead of a second whole one)
  - Job search guys have 2.5 of job search, so don't get 2nd day off
* `&`
  - Describe on master schedule as: `'<span style="background-color:lightcoral">& TABE</span'`
* `^` Light Purple, & caret: class (stays same each week, but edited
  after group's 4 weeks are complete)
  - restricted to regular on-site work day (not lighthouse, not truck, not kitchen since starts 4)
  - Wed caret implies either Communication (" ^" with extra space
    implies Communication) or Computer Class.
    - change to:
      - "^^" Computer Class
      - "^" Communication
  - Tue caret implies GED
  - Fri caret implies Discipleship

### Slot settings (0-1 per slot)
* Dark Purple: Keep day off after overnight desk (don't change)
* Yellow don't change (directive from mgmt, filling in for staff, or resident requested day off)

### Person Permissions
* Describe on Master schedule as:
  ```HTML
<span color="gold">&bull;</span> Light Duty<br/>
span color="limegreen">&bull; Kitchen</span><br/>
<span color="steelblue">&bull; TRUCK</span>
```
* Permissions and Colors
  * blue: truck
  * green: kitchen
  * red: staff stand-in ("sub")
  * darkorange: lighthouse

## Work Assignments
* [ ] Handwritten number code on master schedule determines how roster
  is filled (change to `AT_ShowOnDayAs` on day roster), but that is not needed if
  the week schedule is on a computerized display--only needed if
  printed.
* Each roster is either phase 1 (foundation) or phase 2 (career)
  * Foundation: some need GED, some need math. Includes anyone:
    - not employed
    - not student
    - not on job search
    - not career
* Sorting Room is last (It is a catch-all for anyone working but not
  scheduled anywhere else).


### Questions for Master Scheduler
* [x] Is there ever a day off that is changeable but not split?
  - If there is nothing else that day, that is how OFF is distinguish
  - Fill all that must be off, then save rest for the end--avoid too
    many people off at ones
* [x] What is TABE?
  - It is a test every monday afternoon (Mon 1-3).
* [x] Regarding "not guys in purple": Where is purple shown?
  - Dark purple: after overnight, ensure that day remains off
  - add flag that prevents truck/lighthouse as separate issue
    - [x] Should that be hidden from master schedule? Yes.
* [x] One ID is red:--why? (reason is unknown, so I added
  a P_Note column to People as a basis for highlighting the ID and
  adding a message.

* [ ] Phase corresponding to...Counseling, Tutoring, Computer Class, Other Class
* [ ] Why is JobSearch sometimes red?
* [ ] Why is JobSearch sometimes italicized?

### First-time setup only
#### Oracle Only
* Run the setup-db-oracle-xe-win.sql file IF on Windows
  - You must first modify the SQL file if your pluggable database seed
    is not in:
    `C:\APP\OWNER\PRODUCT\18.0.0\ORADATA\XE\pdbseed`
    (after changing the two paths, the script should work on your
    operating system and version)
* Run the setup-tables.sql file
* OPTIONALLY run the setup-sample-data.sql file (only for demonstration
  purposes).
* Connect to the database and change the password. The script above sets
  the "pbappsetup" password to "arst2020#PW".
  - Example connection using SQL Developer:
    - username: 
    - password: 
    - role: SYSDBA
    - Connection Type: Local/Bequeeth
    (Data Science Center, 2017)
  - list SIDs using Command Prompt (not SQL Plus!) like this:
    `lsnrctl status`
* Connecting normally
(after running setup-db-oracle-xe-win.sql)
  - Enter user-name: `connect pbappsetup/arst2020#PW@//localhost:1521/wdpdb` WORKS
  - `connect pbappsetup/arst2020#PW@//localhost:1521/wdpdb` WORKS
  - Enter user-name: `pbappsetup/arst2020#PW@wdpdb`
  - `connect pbappsetup/arst2020#PW@wdpdb`

#### universal steps 
* get npm packages
```
npm install
```

#### First-time development only
```
#npm init
#npm install oracledb express cookie-parser body-parser ejs
npm install express-generator -g
express --view=ejs planbranch --git
# git: option adds .gitignore
# --view=ejs (formerly --ejs): adds ejs engine support
cd planbranch
npm install
npm install oracledb install cookie-parser body-parser
```

## Run
* run with debugging: `SET DEBUG=planbranch:* & npm start`


## References
Data Science Center. (n.d.). *How to connect SQL DEVELOPER ORACLE 12C: ORA-01017: invalid user name/password; logon denied*. Retrieved from https://www.youtube.com/watch?v=0qh6NG3UcgM 


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a
Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased] - 2019-06-24
### Added
- Implement forms and other partials.
- Create icon.
- Hard-code some of the sample data.
- Add table definitions.
- Add `alter` statements temporarily to upgrade old database
  (see docs/setup-tables.sql).
- Create sequence for `ScheduleItems` primary key.

### Changed
- Manually specify primary key value so that sequence and trigger are
  not needed (so, for example, testing as Oracle SYSDBA is possible).
- Rename db user from appwd and appwdsetup to pbapp and pbappsetup.
- Change `time` to `char(5)` (fix PL/SQL syntax).
- Refererence Weeks not Week (fix bug).
- Correct use of precision (first param counts all digits, including
  count specified as second param).
- Increase string sizes to account for sample data.
- Change `number(22)` to `varchar2(22)` (fix bug).
- Change foreign key PP_Id in People to `integer` (allow relationship).
- Add `not null` as necessary to allow setting `primary key`.


functions.js


var dataFields = {};

exports.defineTable = function (tableName, obj) {
	dataFields[tableName] = obj;
}

var sampleData = {};

exports.getFormSpecs = function () {
	return dataFields;
}

exports.getSampleData = function () {
	return sampleData;
}

exports.pushSampleData = function (tableName, valueList) {
	sampleData[tableName] = {};
	var dstLength = dataFields[tableName].length;
	var srcLength = valueList.length;
	var fields = dataFields[tableName];
	var i;
	for (i = 0; i < dstLength; i++) {
		if (i > srcLength) {
			break;
		}
		sampleData[tableName][fields[i].name] = valueList[i];
	}
	if (i < dstLength) {
		console.error(i + " < " + tableName + ".length for: ");
		console.error("  " + valueList.join(', '));
	}
}


exports.restrict = function (req, res, next) {
	//see <http://toon.io/on-passportjs-specific-use-cases/>
	//express-specific improvements from <https://scotch.io/tutorials/route-middleware-to-check-if-a-user-is-authenticated-in-node-js>
	//if (req.isUnAuthenticated())
	return next;
	// TODO: authenticate instead:
	// if (!req.user || !req.user.username) {
	// 	//only Angular (?) return req.json(403, {message: 'Access denied, please log in'});
	// 	var results = {"error":"Access denied"};
	// 	res.send(results);
	// }
    // else return next();
}


routes/index.js
var express = require('express');
var router = express.Router();
fun = require('../functions.js');

/* GET home page. */
router.get('/', function(req, res, next) {
  var ret = {};
  res.render('index', ret);
});

router.get('/people', function(req, res, next) {
  var ret = {
    formSpecs: fun.getFormSpecs(),
    tables: fun.getSampleData(),
  };
  res.render('people', ret);
});

router.get('/activities', function(req, res, next) {
  var ret = {
    formSpecs: fun.getFormSpecs(),
    tables: fun.getSampleData(),
  };
  res.render('activities', ret);
});

// Use the restrict middleware defined in functions.js as per <http://toon.io/on-passportjs-specific-use-cases/>
router.get('/activitiesjson', fun.restrict, function(req, res, next) {
	res.setHeader("content-type", "application/json; charset=utf-8");
	// MUST call res.send or call multiple res.write then one res.end:
	fun.listEntriesJSON(req, res, "ActivitiesTypes");
});

router.get('/scheduleitem', function(req, res, next) {
  var fwd = ["P_Id", "WSC_Id", "W_Id"];
  var ret = {
    errStr: null,
    formSpecs: fun.getFormSpecs(),
    tables: fun.getSampleData(),
  };
  var fwdLength = fwd.length;
  for (var i = 0; i < fwdLength; i++) {
    if (typeof req.query[fwd[i]] === 'undefined') {
      if (ret.errStr===null) ret.errStr="";
      ret.errStr+="Missing "+fwd[i]+". ";
      ret[fwd[i]] = null;
    }
    else ret[fwd[i]] = req.query[fwd[i]];
  }
  res.render('scheduleitem', ret);
});

module.exports = router;


public/javascripts/main.js



var timePickerData = [['7 AM','07:00'],['8 AM','08:00'],['9 AM','09:00'],['10 AM','10:00'],['11 AM','11:00'],['12 PM','12:00'],['1 PM','13:00'],['2 PM','14:00'],['3 PM','15:00'],['4 PM','16:00'],['5 PM','17:OO'],['6 PM','18:00']];

function generateDropDownFromPairs(table, nameStr, idStr, classStr) {
    var ret = "";
    var inLength = pairs.length;
    for (var i = 0; i < inLength; i++) {
        var caption = table[i][nameStr];
        ret += "\n        <option value=\""+table[i][nameStr]+'">'+caption+'</option>'
    }
    if (ret.length > 0) {
        var propertiesStr = "";
        if (typeof classStr !== 'undefined')
            propertiesStr += ' class="'+classStr+'"';
        if (typeof idStr !== 'undefined')
            propertiesStr += ' id="'+idStr+'"';
        if (typeof nameStr !== 'undefined')
            propertiesStr += ' name="'+nameStr+'"';
        
        ret = "\n      <select"+propertiesStr+">\n" + ret + "\n      </select>\n";
    }
    return ret;
}

// TODO: maxLength
function generateDropDownFromPairs(pairs, nameStr, idStr, classStr) {
    var ret = "";
    var inLength = pairs.length;
    for (var i = 0; i < inLength; i++) {
        ret += "\n        <option value=\""+pairs[i][1]+'">'+pairs[i][0]+'</option>'
    }
    if (ret.length > 0) {
        var propertiesStr = "";
        if (typeof classStr !== 'undefined')
            propertiesStr += ' class="'+classStr+'"';
        if (typeof idStr !== 'undefined')
            propertiesStr += ' id="'+idStr+'"';
        if (typeof nameStr !== 'undefined')
            propertiesStr += ' name="'+nameStr+'"';
        
        ret = "\n      <select"+propertiesStr+">\n" + ret + "\n      </select>\n";
    }
    return ret;
}

function generateForm(formSpec, tables) {
    var ret = '<form action="/" method="get">';
    var inLength = formSpec.length;
    for (var i = 0; i < inLength; i++) {
        
        var field = formSpec[i];
        var caption = field.name;
        var fromTable = false;
        var fieldHtml = "";
        if (typeof field.sourceTable !== 'undefined') {
            fromTable = true;
        }
        if (typeof field.caption !== 'undefined') {
            caption = field.caption;
        }
        else {
            if (!fromTable) {
                caption = caption.substring(caption.indexOf("_") + 1);
            }
            caption = caption.replace("_", " ");
        }
        if (fromTable) {
            if (typeof tables[field.sourceTable] === 'object') {
                fieldHtml = generateDropDown(tables[field.sourceTable], field.name, field.name, "form-control");
                if (fieldHtml.length > 0) {
                    fieldHtml = "\n    "+'<div class="form-group">'
                        +"\n      "+'<label for="'+field.name+'">'+caption+'</label>';
                        + fieldHtml
                        +"\n    "+'</div>';
                }
            }
        }
        if (fieldHtml.length > 0) {
            ret += fieldHtml;
        }
        else if (field.inputType == 'timePicker') {
            ret += "\n    "+'<div class="form-group">'
                 + "\n      "+'<label for="'+field.name+'">'+caption+'</label>';
                 + generateDropDownFromPairs(timePickerData, field.name, field.name, "form-control");
                 + "\n    "+'</div>';
        }
        else if (field.inputType == 'checkbox') {
            ret += "\n    "+'<div class="form-group form-check">'
                    + "\n      "+'<input type="checkbox" class="form-check-input" name="'+field.name+'" id="'+field.name+'">'
                    + "\n      "+'<label class="form-check-label" for="'+field.name+'">'+caption+'</label>'
                    + "\n    "+'</div>';
        }
        else if (field.inputType == 'textarea' || field.inputType == 'memo') {
            // ret += '<textarea name="'+field.name+'" rows="4" cols="50"></textarea>';
            ret += "\n    "+'<div class="form-group">'
                    + "\n      "+'<label for="'+field.name+'">Example textarea</label>'
                    + "\n      "+'<textarea class="form-control" name="'+field.name+'" id="'+field.name+'" rows="3"></textarea>'
                    + "\n    "+'</div>';
        }
        else {
            ret += "\n" + '<input name="'+field.name+'" id="'+field.name+'" class="form-control" type="text" placeholder="'+caption+'">';
        }
    }
    ret += '<button type="submit" class="btn btn-primary">Submit</button>';
    ret += "\n</form>";
    return ret;
}

app.js
var createError = require('http-errors');
var express = require('express');
var path = require('path');
var cookieParser = require('cookie-parser');
var logger = require('morgan');

var indexRouter = require('./routes/index');
var usersRouter = require('./routes/users');

var app = express();

// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'ejs');

app.use(logger('dev'));
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
//app.use(express.static(path.join(__dirname, 'js'))); //don't do this or it'll be root of website-- use "/javascripts" to access public/javascripts instead
app.use(express.static(path.join(__dirname, 'public')));

app.use('/', indexRouter);
app.use('/users', usersRouter);

// catch 404 and forward to error handler
app.use(function(req, res, next) {
  next(createError(404));
});

// error handler
app.use(function(err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.render('error');
});

module.exports = app;

// Default input: text box
// TODO: annotate child tables that use value
// TODO: Default Caption: remove anything before first underscore, replace underscores with spaces
// TODO: if using a NoSQL db, don't save false booleans nor empty strings
fun.defineTable("ActivitiesTypes", [
  {name:'AT_Id', type:'integer', generated:true},
  {name:'WPT_Id', type:'integer', caption:'Required Permission', sourceTable:'WorkPermissionsTypes'},
  {name:'PP_Id', type:'integer', caption:'Phase'},
  {name:'AT_ShowOnDay', default:false},
  {name:'AT_ShowOnDayAs', inputType:'memo', caption:'Description', required:true},
  {name:'AT_ShowOnWeekAs'},
  {name:'AT_Bold', type:'boolean', inputType:'checkbox', default:false},
  {name:'AT_Italics', type:'boolean', inputType:'checkbox', default:false},
  {name:'AT_FGColor', type:'string', inputType:'colorpicker'},
  {name:'AT_BGColor', type:'string', inputType:'colorpicker'},
  {name:'AT_TimeOnWeekAs', type:'string', caption:"Time Format on Weekly"},
  {name:'AT_TimeColor', type:'string', inputType:'colorPicker'},
  {name:'AT_LimitPerWeek', type:'integer'},
  {name:'AT_FractionOfDay', type:'float', required:true, default:0.5},
  {name:'AT_Available', type:'boolean', inputType:'checkbox', default:true},
  {name:'SIR_Id', type:'integer', sourceTable:'ScheduleItemsReasons', required:false, tip:'Only define this if you want to pre-fill the reason!'},
]);

fun.defineTable("PeoplePhases", [
  {name:'PP_Id', type:'integer'},
]);

fun.defineTable("People", [
  {name:'P_Id', generated:true},
  {name:'P_LastName'},
  {name:'P_FirstName'},
  {name:'P_MiddleName'},
  {name:'PP_Id', sourceTable:'PeoplePhases', required:true, default:1},
  {name:'PP_Active', default:1, inputType:'checkbox'},
  {name:'PP_Note'},
]);


// TODO: implement values list values:[{'1':'13:00'},{'2':'14:00'}]

//sourceTable is commented since context prefills value
fun.defineTable("ScheduleItems", [
  {name:'SI_Id', generated:true},
  {name:'P_Id', required:true}, //sourceTable:'People'
  {name:'W_Id', required:true}, //sourceTable:'Weeks'
  {name:'WSC_Id', required:true}, //sourceTable:'WeekScheduleColumns'
  {name:'AT_Id', type:'integer', sourceTable:'ActivitiesTypes', required:true},
  {name:'SI_Time', type:'time', inputType:'timePicker', caption:'Time (HH:MM)'},
  {name:'SI_ShowOnWkAs'},
  {name:'SIR_Id', sourceTable:'ScheduleItemsReasons', caption: "Reason ID"},
]);